### PR TITLE
fix(ec2) Add Ingress rule for VPC Endpoints & CIDR Blocks

### PR DESCRIPTION
**Pull request by jd-carroll**
* Add ingress rule for VPC Endpoints (#4589)
* Expose CIDR block information on `IVpc` (#2232)

### CODE REVIEW

**Comment by jogold** [comment]
I'm not sure that allowing traffic from all addresses in the VPC CIDR is good default here (security-wise).

One must use the `connections` object, as originally documented:
https://github.com/aws/aws-cdk/blob/0bce4d694155422a0a807096788fecc1f0683c33/packages/%40aws-cdk/aws-ec2/test/integ.vpc-endpoint.lit.ts#L38-L41

**Comment by jd-carroll** [comment]
@jogold Can you provide an example where that is a security risk?  I think the code you just referenced is more of a security risk than using the VPC CIDR, and I can't think of a scenario where I would want some resources in my VPC to use a private endpoint but others not.

The code changes here are more in-line with AWS documentation, such as:
https://aws.amazon.com/premiumsupport/knowledge-center/ec2-systems-manager-vpc-endpoints/

**Comment by jogold** [comment]
> @jogold Can you provide an example where that is a security risk? I think the code you just referenced is more of a security risk than using the VPC CIDR, and I can't think of a scenario where I would want some resources in my VPC to use a private endpoint but others not.

The code I referenced is indeed not best practice but was just to illustrate the usage of the `connections` object.

Regarding the security: VPC endpoints allow you to deploy an application in a VPC completely isolated from the internet but still provide access to AWS services. It could be that in that isolated VPC you want to guarantee that your database can never access an AWS service (even if it has been compromised, bad code, injection) or can access one service but not another. Working with the `connections` object allows you to selectively choose who has access. Maybe adding a `allowFromVpc` prop is the right solution here?

**Comment by jd-carroll** [comment]
@jogold - That is an awesome point, I did not think of that. I don’t know I have enough context on where all (or how) the Connections object is used, so I’m not going to try and fit these changes on to that (unless otherwise encouraged). 

It’s probably fair to close this PR, but I’ll leave it open for now as a means to discuss options for #2232. If there are valid things in here which should be kept, happy to amend the PR as necessary. Otherwise it can just be closed.

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 1]
```diff
@@ -938,9 +971,12 @@ export class Vpc extends VpcBase {
 
     this.vpcDefaultNetworkAcl = this.resource.attrDefaultNetworkAcl;
     this.vpcCidrBlockAssociations = this.resource.attrCidrBlockAssociations;
-    this.vpcCidrBlock = this.resource.attrCidrBlock;
+    this.vpcCidrBlock = cidrBlock;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]: Why this change?  What is wrong with `attrCidrBlock`?

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts [hunk 2]
```diff
@@ -363,6 +364,17 @@ export class InterfaceVpcEndpoint extends VpcEndpoint implements IInterfaceVpcEn
     const securityGroup = new SecurityGroup(this, 'SecurityGroup', {
       vpc: props.vpc
     });
+
+    if (props.vpc.vpcCidrBlock) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: I do not feel comfortable with doing this outright. I would feel better if it's behind a flag (suggested name: `allowAllVpcTraffic: true|false`).

It would be nicer to implement as:

```ts
this.connections.allowDefaultPortFrom(Peer.ipv4(props.vpc.vpcCidrBlock));
```

But even doing this only works for securitygroups that have an outbound rule. If not, you would still need to do the following:

```ts
endpoint.connections.allowDefaultPortFrom(myAsg);
```

So is this even worth it?

File: packages/aws-cdk/test/context-providers/test.vpcs.ts [hunk 3]
```diff
@@ -44,6 +44,8 @@ export = nodeunit.testCase({
     // THEN
     test.deepEqual(result, {
       vpcId: 'vpc-1234567',
+      vpcCidrBlock: '10.0.0.0/16',
```
Comment (rix0rrr) [inline packages/aws-cdk/test/context-providers/test.vpcs.ts]: I'm all in favor of this, but be aware there is a concurrent PR changing the VPC provider, and these changes will have to be integrated across PRs: https://github.com/aws/aws-cdk/pull/4544

**Comment by jd-carroll** [comment]
@rix0rrr I generally feel that this PR could be closed, unless there is value related to #2232. I responded to one of your comments above, let me know what your thoughts are on how best to proceed.

**Comment by rix0rrr** [comment]
> I responded to one of your comments above

I'm not seeing it?

**Comment by rix0rrr** [comment]
Closing this until we see a need to reopen.

**Comment by jk21** [comment]
@rix0rrr @jd-carroll with PR #4554 merged now, can this one be progressed? because i don't see the possibility to get the CIDRs of an imported VPC at the moment. (or am i missing sth.?)
